Add presence badge to person avatar

beth-panx opened the pull request:
Closes #268 

### PR Type
- Feature

### Description of the changes
- Please find design [here](https://www.figma.com/file/tVj3DubCqeip6mhw14Up0j/Microsoft-Graph-Toolkit-(main)?node-id=723%3A16751).
- Colors added to mgt-person.css comes from [Fluent UI](https://developer.microsoft.com/en-us/fluentui#/styles/web/colors/presence)
- Presence API [reference docs](https://docs.microsoft.com/en-us/graph/api/presence-get?view=graph-rest-beta&tabs=http).

### How to test:
You can run storybook and see the list of status in [Person Presence Display All](http://localhost:6006/?path=/story/components-mgt-person--person-presence-display-all)
<img width="346" alt="Screen Shot 2020-04-21 at 1 56 25 PM" src="https://user-images.githubusercontent.com/7429891/79913101-eae05880-83d7-11ea-8c41-0c19e43a2edf.png">

You can also `npm start` and replace `body` with the following in `index.html` file
```
  <body>
    <mgt-teams-provider
      client-id="a974dfa0-9f57-49b9-95db-90f04ce2111a"
      auth-popup-url="auth.html"
    ></mgt-teams-provider>

    <mgt-msal-provider
      client-id="a974dfa0-9f57-49b9-95db-90f04ce2111a"
      scopes="user.read,people.read,user.readbasic.all,contacts.read,calendars.read,presence.read"
      depends-on="mgt-teams-provider"
    ></mgt-msal-provider>

    <!-- <mgt-mock-provider></mgt-mock-provider> -->

    <mgt-login></mgt-login>
    <mgt-person person-query="me" show-name show-email person-card="hover" show-presence></mgt-person>
    <mgt-agenda></mgt-agenda>
    <mgt-people
      user-ids="91ca32bf-43f4-4907-8ad5-bec920fb4ad3,a2e824c2-acb9-456b-b159-4c25fdd3c472,0cd426de-cb7a-41df-a51b-d6d189fafcd2,a295d82d-c88e-4f11-a5b3-5e1f7b0da018"
      show-presence
    ></mgt-people>
  </body>
```
### PR checklist
- [x] Project builds (`npm run build`) and changes have been tested in supported browsers (including IE11)
- [x] All public classes and methods have been documented
- [x] Added appropriate [documentation](https://github.com/microsoftgraph/microsoft-graph-docs/tree/master/concepts/toolkit) [target branch `mgt/next` for new features]. Docs PR: https://github.com/microsoftgraph/microsoft-graph-docs/pull/7857
- [x] License header has been added to all new source files (`npm run setLicense`)
- [x] Contains **NO** breaking changes

nmetulev commented on src/components/mgt-person/mgt-person.ts:
```diff
@@ -324,9 +341,154 @@ export class MgtPerson extends MgtTemplatedComponent {
+    if (statusClass === 'presence-oof-offline') {
+      iconHtml = html`
+        <div class="ms-Icon presence-basic presence-oof-offline-wrapper">
+          <svg class="presence-oof-offline" viewBox="0 0 12 10" fill="none" xmlns="http://www.w3.org/2000/svg">
```
You can move this with all the other svgs: https://github.com/microsoftgraph/microsoft-graph-toolkit/blob/master/src/utils/SvgHelper.ts

beth-panx replied:
hmmm can you add class to it?

nmetulev commented on src/components/mgt-person/mgt-person.ts:
```diff
@@ -457,6 +619,30 @@ export class MgtPerson extends MgtTemplatedComponent {
+    };
+
+    // populate presence
+    if (!this.personPresence) {
```
Should this also check for `showPresence` to avoid making unnecessary calls?

nmetulev commented on src/graph/graph.presence.ts:
```diff
@@ -0,0 +1,84 @@
+ * @returns {Promise<Presence>}
+ * @memberof BetaGraph
+ */
+export async function getPersonPresence(graph: IGraph, userId: string): Promise<Presence> {
```
rename to `getUserPresence` since this only works with users

nmetulev commented on src/graph/graph.presence.ts:
```diff
@@ -0,0 +1,84 @@
+ * @returns {}
+ * @memberof BetaGraph
+ */
+export async function getPeoplePresence(graph: IGraph, userIds: string[]) {
```
rename to `getUsersPresence`

nmetulev commented on src/graph/graph.presence.ts:
```diff
@@ -0,0 +1,84 @@
+
+    for (const id of userIds) {
+      // tslint:disable
+      console.log('RESPONSE: ' + JSON.stringify(response));
```
remove console.log

nmetulev commented on src/styles/shared-styles.scss:
```diff
@@ -29,21 +29,27 @@
+  content: '\\\EB6D';
 }
+
+$presence-color-online: #6bb700;
```
I don't see these used outside of the person component, they should stay with the person component.

nmetulev commented on src/components/mgt-people/mgt-people.ts:
```diff
@@ -8,7 +8,9 @@
 import * as MicrosoftGraph from '@microsoft/microsoft-graph-types';
 import { customElement, html, property, TemplateResult } from 'lit-element';
 import { repeat } from 'lit-html/directives/repeat';
+import { BetaGraph } from '../../BetaGraph';
```
Can we avoid importing BetaGraph in people and person by importing it in graph.presence.ts instead?

beth-panx replied:
I think it's awkward at either location... I don't really have a preference. This would be a good @shweaver-MSFT question.

nmetulev commented on src/components/mgt-people/mgt-people.ts:
```diff
@@ -146,6 +158,9 @@ export class MgtPeople extends MgtTemplatedComponent {
   private _peopleQueries: string[];
 
+  // tslint:disable-next-line:completed-docs
+  private _peoplePresence: Array<{ '@odata.context': string; id: string; availability: string; activity: string }>;
```
should this type just be `Presence`?

beth-panx replied:
It's Presence plus the '@data.context' part because of the search.

nmetulev commented on src/components/mgt-people/mgt-people.ts:
```diff
@@ -250,6 +265,17 @@ export class MgtPeople extends MgtTemplatedComponent {
   protected renderPerson(person: MicrosoftGraph.User | MicrosoftGraph.Person | MicrosoftGraph.Contact): TemplateResult {
+    // Get person presence from _peoplePresence to pass to mgt-person
+    let personPresence = null;
+    for (const presenceObj of this._peoplePresence) {
```
You can avoid this for loop by using an object to store the presence by the id of the user

beth-panx replied:
_peoplePresence would still be an array of object though. Are you suggesting using Array.filter() or find()?

nmetulev commented on src/components/mgt-person/mgt-person.ts:
```diff
@@ -457,6 +619,30 @@ export class MgtPerson extends MgtTemplatedComponent {
+    // populate presence
+    if (!this.personPresence) {
+      try {
+        if (this.userId) {
```
We should be able to get presence even if using person query - similar to how we get the image. The presence can also be batched together with the image call for better performance.

beth-panx replied:
Yeah looking at the documentation, seemed like only id would work. Tried upn and email, both can be passed in to the {userid} bracket so will update this.

beth-panx replied:
Actually, after some testing... It seems like the json response will always indicate presence unknown if you pass in a query.

nmetulev replied:
I would first fetch the person (with `person-query` or `user-id`, etc). Once you have the person, you can check if that person is a user (if it has upn or check the type). If it's a user, fetch the presence using the id.

beth-panx replied:
Makes sense. All done

nmetulev replied:
I think you did this in the `mgt-people` but I still need to do this here (in `mgt-person`)

nmetulev commented on stories/components/person.stories.js:
```diff
@@ -28,6 +28,123 @@ export const personPhotoOnly = () => html`
+  <mgt-person person-query="me" show-presence show-email show-name></mgt-person>
+`;
+
+export const personPresenceDisplayAll = () => html`
```
Love this story!!!

nmetulev commented on src/components/mgt-people/mgt-people.ts:
```diff
@@ -288,7 +315,16 @@ export class MgtPeople extends MgtTemplatedComponent {
+        const betaGraph = BetaGraph.fromGraph(graph);
+
+        // populate presence for people only if userIds array exist
+        if (this.showPresence && this.userIds) {
```
You should be able to move this after the people are fetched and then get the presence for people - should work for any people, not just for people fetched by user-id

nmetulev commented on src/components/mgt-person/mgt-person.ts:
```diff
@@ -6,15 +6,19 @@
+import { Presence } from '@microsoft/microsoft-graph-types-beta';
 import { customElement, html, property, query, TemplateResult } from 'lit-element';
 import { classMap } from 'lit-html/directives/class-map';
+import { BetaGraph } from '../../BetaGraph';
```
Same here for `BetaGraph` move to graph.presnce.ts and avoid instantiating a 'BetaGraph` here

beth-panx commented:
@nmetulev I added an attribute to mgt-person as an attempt to update the size of avatar without breaking change to existing implementation. Let me know what you think. If this is what we want to go with, I will update docs accordingly. Also addressed getting person presence based on personDetails in mgt-person.

nmetulev commented on src/components/mgt-person/mgt-person.ts:
```diff
@@ -301,13 +301,13 @@ export class MgtPerson extends MgtTemplatedComponent {
 
     const isLarge = this.showEmail && this.showName;
-    const smallAvatar = this.smallAvatar;
+    const isSmallAvatar = this.avatarSize === 'small' ? true : false;
```
You can simplify this:

```suggestion
    const isSmallAvatar = this.avatarSize === 'small';
```

nmetulev commented on src/components/mgt-person/mgt-person.ts:
```diff
@@ -335,11 +335,11 @@ export class MgtPerson extends MgtTemplatedComponent {
 
     const isLarge = this.showEmail && this.showName;
-    const smallAvatar = this.smallAvatar;
+    const isSmallAvatar = this.avatarSize === 'small' ? true : false;
```
same

```suggestion
    const isSmallAvatar = this.avatarSize === 'small';
```

nmetulev commented on src/components/mgt-person/mgt-person.ts:
```diff
@@ -506,11 +506,11 @@ export class MgtPerson extends MgtTemplatedComponent {
         : '';
     const isLarge = this.showEmail && this.showName;
-    const smallAvatar = this.smallAvatar;
+    const isSmallAvatar = this.avatarSize === 'small' ? true : false;
```
same

```suggestion
    const isSmallAvatar = this.avatarSize === 'small';
```

nmetulev commented on src/components/mgt-person/mgt-person.ts:
```diff
@@ -551,10 +551,10 @@ export class MgtPerson extends MgtTemplatedComponent {
     const emailTemplate: TemplateResult = this.showEmail ? this.renderEmail(email) : html``;
     const nameTemplate: TemplateResult = this.showName ? this.renderName(person.displayName) : html``;
-    const smallAvatar = this.smallAvatar;
+    const isSmallAvatar = this.avatarSize === 'small' ? true : false;
```
same

```suggestion
    const isSmallAvatar = this.avatarSize === 'small';
```
